### PR TITLE
MM-36687, MM-38302, MM-37598 Fix push notifications with CRT

### PR DESCRIPTION
**Pull request by anurag6713**
#### Summary
When CRT is enabled...
* Clears channel notifications and thread notifications on read separately
*  Android now shows "Threads" label instead of "Channel name".
* "X new messages" in the channel now counts only the root posts.

### Related PRs
https://github.com/mattermost/mattermost-server/pull/18347
https://github.com/mattermost/mattermost-push-proxy/pull/85
https://github.com/mattermost/enterprise/pull/1096

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-36687
https://mattermost.atlassian.net/browse/MM-38302
https://mattermost.atlassian.net/browse/MM-37598

#### Device Information
This PR was tested on: iPhone iOS 14.7.1, Android  11

#### Release Note
```release-note
Fixes clearing & grouping of push notifications on iOS & android when CRT is enabled
```

### CODE REVIEW

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 1]
```diff
@@ -262,23 +304,39 @@ private static void saveNotificationsMap(Context context, Map<String, List<Integ
         }
     }
 
-    private static Map<String, List<Integer>> loadNotificationsMap(Context context) {
-        Map<String, List<Integer>> outputMap = new HashMap<>();
+    /**
+     * Map Structure
+     * 
+     * {
+     *     channel_id1: {
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:311]: Why did you changed the structure we discussed?

Comment (anurag6713) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:311]: I still haven't made the changes after the discussion. I was thinking about the use cases with/without CRT (if someone toggles). It's about adding more conditions and I want to see which structure makes it more readable.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 2]
```diff
@@ -61,63 +58,78 @@ public CustomPushNotification(Context context, Bundle bundle, AppLifecycleFacade
...
 
-    public static void clearChannelNotifications(Context context, String channelId) {
+    public static void clearChannelNotifications(Context context, String channelId, String rootId, Boolean isCRTEnabled) {
         if (!android.text.TextUtils.isEmpty(channelId)) {
-            Map<String, List<Integer>> notificationsInChannel = loadNotificationsMap(context);
-            List<Integer> notifications = notificationsInChannel.get(channelId);
+            final NotificationManagerCompat notificationManager = NotificationManagerCompat.from(context);
+
+            // rootId is available only when CRT is enabled & clearing the thread
+            final boolean isClearThread = !android.text.TextUtils.isEmpty(rootId);
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:112]:
```suggestion
            final boolean isClearThread = isCRTEnabled && !android.text.TextUtils.isEmpty(rootId);
```

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 3]
```diff
@@ -132,12 +144,12 @@ public void onReceived() {
         final String ackId = initialData.getString("ack_id");
         final String postId = initialData.getString("post_id");
         final String channelId = initialData.getString("channel_id");
+        final String rootId = initialData.getString("root_id");
+        final boolean isCRTEnabled = initialData.getString("is_crt_enabled") != null && initialData.getString("is_crt_enabled").equals("true");
         final boolean isIdLoaded = initialData.getString("id_loaded") != null && initialData.getString("id_loaded").equals("true");
         int notificationId = CustomPushNotificationHelper.MESSAGE_NOTIFICATION_ID;
         if (postId != null) {
             notificationId = postId.hashCode();
-        } else if (channelId != null) {
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:139]: Why did you remove this one ?

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java [hunk 4]
```diff
@@ -165,32 +177,49 @@ public void reject(String code, String message) {
...
+                                    post.put("post_id", postId);
+                                }
+
+                                final Boolean isThreadNotification = isCRTEnabled && post.has("root_id");
+                                final String groupId = isThreadNotification ? rootId : channelId;
+                                
+                                Map<String, Map<String, JSONObject>> notificationsInChannel = loadNotificationsMap(mContext);
+                                Map<String, JSONObject> notifications = notificationsInChannel.get(groupId);
+                                if (notifications == null) {
+                                    notifications = new HashMap<String, JSONObject>();
```
Comment (enahum) [inline android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:196]: shouldn't this be synchronized ?

File: app/constants/view.js [hunk 5]
```diff
@@ -118,7 +118,7 @@ export default {
     ...ViewTypes,
     RequiredServer,
     POST_VISIBILITY_CHUNK_SIZE: 60,
-    CRT_CHUNK_SIZE: 60,
+    CRT_CHUNK_SIZE: 30,
```
Comment (enahum) [inline app/constants/view.js:121]: Why 30 instead of 60 ?

File: app/init/push_notifications.ts [hunk 6]
```diff
@@ -75,34 +84,56 @@ class PushNotifications {
...
+            const notification = notifications[i];
+            if (notification.channel_id === channelId) {
+                let doesNotificationMatch: boolean;
+                if (clearThreads) {
+                    doesNotificationMatch = notification.root_id === rootId;
+                } else if (collapsedThreadsEnabled) {
+                    // Do not match when CRT is enabled BUT post is not a root post
+                    doesNotificationMatch = !notification.root_id;
+                } else {
+                    doesNotificationMatch = true;
```
Comment (enahum) [inline app/init/push_notifications.ts:108]: you can set this as the default value and then remove this else statement

File: app/screens/channel/channel.android.js [hunk 7]
```diff
@@ -98,6 +98,7 @@ export default class ChannelAndroid extends ChannelBase {
                     openMainSidebar={this.openMainSidebar}
                     openSettingsSidebar={this.openSettingsSidebar}
                     onPress={this.goToChannelInfo}
+                    isGlobalThreads={viewingGlobalThreads}
```
Comment (enahum) [inline app/screens/channel/channel.android.js:101]: was this missing before ?

File: ios/Mattermost/AppDelegate.m [hunk 8]
```diff
@@ -111,9 +113,25 @@ -(void)cleanNotificationsFromChannel:(NSString *)channelId {
...
+            // Check if it is a thread notification
+            if (rootId != nil) {
+              doesNotificationMatch = [pId isEqualToString: rootId] || [rId isEqualToString: rootId];
+            } else {
+              // With CRT ON, remove notifications without rootId
+              doesNotificationMatch = rId == nil;
+            }
+          } else {
+            // Default condition, without CRT
+            doesNotificationMatch = true;
```
Comment (enahum) [inline ios/Mattermost/AppDelegate.m:130]: same comment as in the JS side

**Review by enahum — CHANGES_REQUESTED** [review]
Also I think you should be setting `is_crt` value as boolean for both platforms in the push proxy

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 3 above)
Comment (anurag6713) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:139]: When we went through the code, we discussed about the possible use cases where postId is not available and decided to remove this one.

Comment (enahum) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:139]: Have you tried on Android the "Send test notification" and validate this does not cause any issues ?

File: app/constants/view.js (hunk 5 above)
Comment (anurag6713) [reply app/constants/view.js:121]: Sorry this was included in this PR. Web app has implemented infinite scrolling and we had a discussion about number of threads to load and arrived at this number.

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java (hunk 3 above)
Comment (anurag6713) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:139]: Yes it is working. Can I revert the code in case we are unable to recall the edge case where we might need this?

Comment (enahum) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.java:139]: Yeah, I don;t really recall which is the case this was needed, perhaps is safest to keep

File: app/screens/channel/channel.android.js (hunk 7 above)
Comment (anurag6713) [reply app/screens/channel/channel.android.js:101]: Yes ... this was missing before.

**Review by enahum — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @anurag6713 
I ran into few issues while testing
I used https://v5351.test.mattermost.cloud pointed to server PR commit, due to issue with test servers failing to create on the server PR, updated push notification server manually with https://push-test.mattermost.com
1. Not receiving any notifications on the iOS app built for the PR + actual device (iPhone Xs Max v14.8)
2. Android vm - I was receiving notifications and they would clear from the locked screen after I read the channel on a different device, but the app became unusable after I replied to notification received on locked screen. I don't have an actual Android device to re-test this and rule out if it's vm related.
AndroidVD info:
Name: Pixel_5_API_30
CPU/ABI: Google APIs Intel Atom (x86)
Screen Shot 2021-10-15 at 3 46 10 PM

**Comment by enahum** [comment]
@anurag6713 can you please resolve the conflicts on this ASAP!?>

**Comment by enahum** [comment]
@amyblais why was this pushed to 1.49 ?

**Comment by amyblais** [comment]
> @amyblais why was this pushed to 1.49 ?

@enahum It could be included in v1.48, but the server/enterprise PRs are still open and I'm not sure if they have to go in at the same time.

**Comment by enahum** [comment]
@amyblais thanks for the context.
